catalogd/internal/storage/localdir.go

```diff
@@ -20,95 +26,266 @@ import (
 // done so that clients accessing the content stored in RootDir/catalogName have
 // atomic view of the content for a catalog.
 type LocalDirV1 struct {
-	RootDir string
-	RootURL *url.URL
+	RootDir            string
+	RootURL            *url.URL
+	EnableQueryHandler bool
+
+	m  sync.RWMutex
+	sf singleflight.Group
 }
 
-const (
-	v1ApiPath = "api/v1"
-	v1ApiData = "all"
+var (
+	_                Instance = &LocalDirV1{}
+	ErrInvalidParams          = errors.New("invalid parameters")
 )
 
-func (s LocalDirV1) Store(ctx context.Context, catalog string, fsys fs.FS) error {
-	fbcDir := filepath.Join(s.RootDir, catalog, v1ApiPath)
-	if err := os.MkdirAll(fbcDir, 0700); err != nil {
+func (s *LocalDirV1) Store(ctx context.Context, catalog string, fsys fs.FS) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	if err := os.MkdirAll(s.RootDir, 0700); err != nil {
 		return err
 	}
-	tempFile, err := os.CreateTemp(s.RootDir, fmt.Sprint(catalog))
+	tmpCatalogDir, err := os.MkdirTemp(s.RootDir, fmt.Sprintf(".%s-*", catalog))
 	if err != nil {
 		return err
 	}
-	defer os.Remove(tempFile.Name())
-	if err := declcfg.WalkMetasFS(ctx, fsys, func(path string, meta *declcfg.Meta, err error) error {
+	defer os.RemoveAll(tmpCatalogDir)
+
+	storeMetaFuncs := []storeMetasFunc{storeCatalogData}
+	if s.EnableQueryHandler {
+		storeMetaFuncs = append(storeMetaFuncs, storeIndexData)
+	}
+
+	var (
```

Comment:
Throughout these lines where we setup and run concurrent channel writers and readers, we should add some comments to explain what is happening.